module5/lessons/Week1/RefactoringPatterns.md

```diff
@@ -0,0 +1,415 @@
+And now our Slide Player is just a mess.
+
+There are really only 3 common methods for all of these players, so it would make the most sense to break up our IMediaPlayer into 3 seperate interfaces:
+
+```c#
+public interface IPlayable
+{
+    void Play();
+    void Pause();
+    void Stop();
+}
+
+public interface IVolumeAdjustable
+{
+    void AdjustVolume();
+}
+
+public interface ISubtitleSupport
+{
+    void ShowSubtitle();
+}
+
+public class VideoPlayer : IPlayable, IVolumeAdjustable, ISubtitleSupport
+{
+    // Implement relevant methods
+}
+
+public class AudioPlayer : IPlayable, IVolumeAdjustable
+{
+    // Implement relevant methods
+}
+
+public class SlidePlayer : IPlayable
+{
+    // Implement relevant methods
+}
+```
+
+With the **Extract Interfaces** method, we have smaller, more re-usable interfaces!  This also adheres better to the Interface Segregation principle.
+
+## Move Method
+
+**The Problem**: A class has too many responsibilities.  Typically, this happens when a method on a class depends more on _another_ class than on the class it is housed in.  Take this example:
+
+```c#
+class AddressDetails
+{
+    internal string Street { get; set; }
+    internal int Number { get; set; }
+    internal string City { get; set; }
+    internal string PostalCode { get; set; }
+    internal string Country { get; set; }
+}
+
+class Person
+{
+    public Person()
+    {
+        Address = new AddressDetails();
+    }
+
+    public AddressDetails Address { get; set; }
+
+    internal string Name { get; set; }
+    internal string TelephoneNumber { get; set; }
+    internal string Email { get; set; }
+
+    internal string GetFullAddress()
+    {
+        return $"{Address.Street} {Address.Number}, {Address.PostalCode} {Address.City}, {Address.Country}";
+    }
+}
+```
+
+In this Person class, there is a method that relies very heavily on the Address class.  The code works as-is, but it could be refactored to better adhere to the **single responsibility principle**:
+
+```c#
+class AddressDetails
+{
+    internal string Street { get; set; }
+    internal int Number { get; set; }
+    internal string City { get; set; }
+    internal string PostalCode { get; set; }
+    internal string Country { get; set; }
+
+    internal string GetFormattedAddress()
+    {
+        return $"{Street} {Number}, {PostalCode} {City}, {Country}";
+    }
+}
+
+class Person
+{
+    public Person()
+    {
+        Address = new AddressDetails();
+    }
+
+    public AddressDetails Address { get; set; }
+
+    internal string Name { get; set; }
+    internal string TelephoneNumber { get; set; }
+    internal string Email { get; set; }
+
+    internal string GetFullAddress()
+    {
+        return Address.GetFormattedAddress;
+    }
+}
+```
+
+In this refactored version, we are not changing how we interact with the person class (so we know all of our existing code will still run); but we are changing _where_ the logic of getting a formatted address string lives.  
+
```

Comment:
It feels a bit off to me to have a refactoring lesson and not talk about DRY ( Don't Repeat Yourself). Seeing a bunch of repeated code and then breaking that out into a method seems like one of the most common refactoring patterns for our students. I do like your other three examples too though! Maybe we add a forth here?

Reply:
It also feels like this is sort of a subset of Refact Method. Like you should definitely definitely extract method if there is repeated code. Maybe after you introduce Extract Method and give the above example there could be a mini exercise where they try it in a situation with repeated code that needs extracting?